### ISSUE TITLE
speedup of sum(x**i/i,i=1..400)

### ISSUE THREAD

**Issue by certik** [sympy/sympy#3782]
bc.. Currently, 

sum(x**i/i,i=1..400)

is slow like hell, compared to sympycore (300x). The result of this simple
script:
# ! /usr/bin/python2.5

import sympy
from time import time as clock
import pstats
import cProfile
profile = cProfile

def test_summation_performance():
    """`sum(x**i/i,i=1..400)`"""
    s = getattr(sympy,'Integer', int)(0)
    x = sympy.Symbol('x')
    i = 401
    t1 = clock()
    while i:
        s += x**i/i
        i -= 1
    t2 = clock()
    return t2-t1

p = profile.Profile()
p.run("test_summation_performance()")

s = pstats.Stats(p)
s.sort_stats("time", "name").print_stats()

is

```
     13101089 function calls (12317776 primitive calls) in 50.432 CPU
```

seconds

   Ordered by: internal time, function name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
1066896/533852   23.641    0.000   41.205    0.000
/home/ondra/ext/sympy/sympy/core/basic.py:239(compare)
3200664/3199000    4.378    0.000    4.402    0.000 {cmp}
  1066841    3.509    0.000    3.509    0.000
/home/ondra/ext/sympy/sympy/core/numbers.py:439(_hashable_content)
  1960009    3.096    0.000    3.096    0.000 {isinstance}
  2142975    2.840    0.000    2.840    0.000 {len}
      401    2.416    0.006   48.946    0.122
/home/ondra/ext/sympy/sympy/core/add.py:15(flatten)
  1067687    2.297    0.000    2.297    0.000
/home/ondra/ext/sympy/sympy/core/basic.py:227(_hashable_content)
  1067266    2.099    0.000    2.099    0.000 {zip}
498023/415424    1.147    0.000    1.365    0.000
/home/ondra/ext/sympy/sympy/core/basic.py:218(__hash__)
     3603    1.058    0.000   42.263    0.012 {method 'sort' of 'list' objects}
163604/1604    0.842    0.000   50.090    0.031
/home/ondra/ext/sympy/sympy/core/cache.py:88(wrapper)
85407/84205    0.469    0.000    1.013    0.000
/home/ondra/ext/sympy/sympy/core/basic.py:199(**getattr**)
   161599    0.372    0.000    0.372    0.000
/home/ondra/ext/sympy/sympy/core/basic.py:490(args)
83006/81805    0.363    0.000    0.414    0.000
/home/ondra/ext/sympy/sympy/core/assumptions.py:133(_get_assumption)
   124622    0.318    0.000    0.489    0.000 {method 'has_key' of 'dict'
objects}
     2801    0.258    0.000    0.556    0.000
/home/ondra/ext/sympy/sympy/core/assumptions.py:183(assume)
    96605    0.158    0.000    0.158    0.000 {method 'startswith' of 'str'
objects}
    84205    0.157    0.000    0.157    0.000 {method 'pop' of 'list' objects}

Maybe the profiler is lying, but if it is not, the majority of time is
spent in:

```
def compare(self, other):
    """
    Return -1,0,1 if the object is smaller, equal, or greater than other
    (not always in mathematical sense).
    If the object is of different type from other then their classes
    are ordered according to sorted_classes list.
    """
    # all redefinitions of __cmp__ method should start with the
    # following three lines:
    if self is other: return 0
    c = cmp(self.__class__, other.__class__)
    if c: return c
    #
    st = self._hashable_content()
    ot = other._hashable_content()
    c = cmp(len(st),len(ot))
    if c: return c
    for l,r in zip(st,ot):
        if isinstance(l, Basic):
            c = l.compare(r)
        else:
            c = cmp(l, r)
        if c: return c
```

Basically 42s is spent in list.sort(), which basically just calls compare.
So the fix should imho be quite easy -- efficient comparisons. What do you
think Pearu? I am sure you have a lot of experience with this. :)

p. Original issue for "#3782":https://github.com/sympy/sympy/issues/3782: "http://code.google.com/p/sympy/issues/detail?id=683":http://code.google.com/p/sympy/issues/detail?id=683

p. Original author: "https://code.google.com/u/104039945248245758823/":https://code.google.com/u/104039945248245758823/

**Comment by oscarbenjamin** [comment]
What is it specifically that should be made faster in this issue?

**Comment by asmeurer** [comment]
I guess

```py
    s = sympy.Integer(0)
    x = sympy.Symbol('x')
    i = 401
    while i:
        s += x**i/i
        i -= 1
```

According to the OP, it used to take seconds (now it runs in under a second). I'm sure it could still be faster, especially considering we don't actually optimize successive adds particularly well.

**Comment by oscarbenjamin** [comment]
The problem is calling `Add.flatten` for each `+`. Partly it's about sorting but it's also just about the fact that `Add.flatten` has to deconstruct and recombine the args each time. The real problem is just automatic evaluation. If `Add.flatten()` wasn't such a mess of special cases it would be possible to optimise it as is to some extent. Micro-optimisations are possible this can't be fixed without removing automatic evaluation from`+` because that makes the repeated add loop inherently quadratic. Unfortunately that also means that `sum(terms)` is quadratic as well.

If Add didn't canonicalise automatically and there was an explicit canonicalisation function then the equivalent of the above slow code would be:
```python
for i in range(401, 0, -1):
    s += x**i/i
    s = canonicalise(s)
```
It would then be obvious that you should move the `canonicalise` call outside of the loop:
```python
for i in range(401, 0, -1):
    s += x**i/i
s = canonicalise(s)
```
As it is you can get the same effect just by collecting a list of terms and calling `Add` once:
```python
terms = [x**i/i for i in range(401, 0, -1)]
s = Add(*terms)
```
